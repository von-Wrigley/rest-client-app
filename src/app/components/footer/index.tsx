"use client";

import { useEffect, useState } from "react";
import Link from "next/link";
import { useTranslations } from "next-intl";
import Skeleton from "../skeleton";
import styles from "./index.module.scss";

const Footer: React.FC = () => {
  const [isLoading, setIsLoading] = useState(true);
  const t = useTranslations("Footer");

  useEffect(() => {
    const timer = setTimeout(() => setIsLoading(false), 2000);
    return () => clearTimeout(timer);
  }, []);

  const githubLinks = [
    { href: "https://github.com/lionna", label: t("github1") },
    { href: "https://github.com/von-Wrigley", label: t("github2") },
    { href: "https://github.com/CROCIATOFAF", label: t("github3") },
  ];

  if (isLoading) {
    return (
      <footer className={styles.footer}>
        <div className={styles.footerContent}>
          <Skeleton variant="medium" className={styles.skeletonLink} />
          <Skeleton variant="medium" className={styles.skeletonLink} />
          <Skeleton variant="medium" className={styles.skeletonLink} />
          <Skeleton variant="small" className={styles.skeletonText} />
          <Skeleton variant="small" className={styles.skeletonImage} />
        </div>
      </footer>
    );
  }

  return (
    <footer className={styles.footer}>
      <div className={styles.footerContent}>
        {githubLinks.map((link, index) => (
          <Link key={index} href={link.href} className={styles.link}>
            {link.label}
          </Link>
        ))}
        <span>&copy; 2025</span>
        <Link
          href="https://rs.school/courses/reactjs"
          className={styles.courseLink}
        >
          <img
            src="/rss-logo.svg"
            alt={t("courseAlt")}
            className={styles.courseLogo}
          />
        </Link>
      </div>
    </footer>
  );
};

export default Footer;
