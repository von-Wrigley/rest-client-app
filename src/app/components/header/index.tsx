"use client";

import { useEffect, useState } from "react";
import Link from "next/link";
import styles from "./index.module.scss";
import LocaleSwitcher from "../locale-switcher";
import { useTranslations } from "next-intl";

const Header: React.FC = () => {
  const [isSticky, setIsSticky] = useState(false);
  const t = useTranslations("Header");

  useEffect(() => {
    const handleScroll = () => {
      setIsSticky(window.scrollY > 50);
    };

    window.addEventListener("scroll", handleScroll);
    return () => window.removeEventListener("scroll", handleScroll);
  }, []);

  return (
    <header
      data-testid="header"
      className={`${styles.header} ${isSticky ? styles.sticky : ""}`}
    >
      <div className={styles.headerContent}>
        <div className={styles.logoContainer}>
          <Link href="/" className={styles.logo}>
            {t("title")}
          </Link>
        </div>
        <div className={styles.headerRight}>
          <img src="/rss-logo.svg" alt="Course Logo" />
          <LocaleSwitcher />
          <div className={styles.authLinks}>
            <Link href="/signin" className={styles.link}>
              {t("signIn")}
            </Link>
            <Link href="/signup" className={styles.link}>
              {t("signUp")}
            </Link>
          </div>
        </div>
      </div>
    </header>
  );
};

export default Header;
