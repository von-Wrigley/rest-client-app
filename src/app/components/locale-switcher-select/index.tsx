"use client";

import { useTransition, useState, useRef, useEffect } from "react";
import { usePathname, useRouter } from "@/i18n/navigation";
import Image from "next/image";
import styles from "./index.module.scss";

type LocaleOption = {
  value: string;
  label: string;
  flagSrc: string;
};

type Props = Readonly<{
  defaultValue: string;
  children: readonly LocaleOption[];
}>;

export default function LocaleSwitcherSelect({
  children,
  defaultValue,
}: Props) {
  const router = useRouter();
  const pathname = usePathname();
  const [isPending, startTransition] = useTransition();

  const [open, setOpen] = useState(false);
  const wrapperRef = useRef<HTMLDivElement>(null);

  const current = children.find((item) => item.value === defaultValue);

  const handleChange = (value: string) => {
    setOpen(false);
    startTransition(() => {
      router.replace({ pathname }, { locale: value });
    });
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        wrapperRef.current &&
        !wrapperRef.current.contains(event.target as Node)
      ) {
        setOpen(false);
      }
    };
    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  }, []);

  return (
    <div className={styles.selectWrapper} ref={wrapperRef}>
      <button
        type="button"
        className={styles.selectButton}
        onClick={() => setOpen((prev) => !prev)}
        disabled={isPending}
      >
        <Image
          src={current?.flagSrc ?? "/flags/en.png"}
          alt={current?.label ?? "Language"}
          width={20}
          height={15}
        />
        <span>{current?.label}</span>
        <span className={styles.arrow} />
      </button>

      {open && (
        <ul className={styles.dropdown}>
          {children.map((item) => (
            <li
              key={item.value}
              className={styles.option}
              onClick={() => handleChange(item.value)}
            >
              <Image
                src={item.flagSrc}
                alt={item.label}
                width={20}
                height={15}
              />
              <span>{item.label}</span>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
}
