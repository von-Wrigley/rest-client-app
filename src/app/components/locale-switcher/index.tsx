"use client";

import { useLocale, useTranslations } from "next-intl";
import { routing } from "@/i18n/routing";
import LocaleSwitcherSelect from "../locale-switcher-select";

export default function LocaleSwitcher() {
  const t = useTranslations("LocaleSwitcher");
  const locale = useLocale();

  return (
    <LocaleSwitcherSelect defaultValue={locale}>
      {routing.locales.map((cur) => ({
        value: cur,
        label: t(`locales.${cur}`),
        flagSrc: `/flags/${cur}.png`,
      }))}
    </LocaleSwitcherSelect>
  );
}
