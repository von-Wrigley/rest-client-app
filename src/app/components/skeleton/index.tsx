"use client";

import React from "react";
import styles from "./index.module.scss";

export interface SkeletonProps {
  variant?: "small" | "medium" | "large";
  className?: string;
}

const Skeleton: React.FC<SkeletonProps> = ({
  variant = "medium",
  className = "",
}) => {
  return (
    <div className={`${styles.skeleton} ${styles[variant]} ${className}`} />
  );
};

export default Skeleton;
