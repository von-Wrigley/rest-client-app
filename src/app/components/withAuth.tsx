"use client";

import { ComponentType, useEffect, useState } from "react";
import { useRouter } from "next/navigation";
import { supabase } from "@/helper/supabaseClient";

type WithAuthProps = {};

function withAuth<P extends WithAuthProps>(WrappedComponent: ComponentType<P>) {
  return function AuthenticatedComponent(props: P) {
    const router = useRouter();
    const [isLoading, setIsLoading] = useState(true);

    useEffect(() => {
      const checkAuth = async () => {
        const {
          data: { session },
        } = await supabase.auth.getSession();

        if (!session) {
          router.push("/unauthorized");
        } else {
          setIsLoading(false);
        }
      };

      checkAuth();
    }, [router]);

    if (isLoading) {
      return (
        <div className="flex justify-center items-center h-screen">
          <p>Loading...</p>
        </div>
      );
    }

    return <WrappedComponent {...props} />;
  };
}

export default withAuth;
