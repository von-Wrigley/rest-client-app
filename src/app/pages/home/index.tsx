"use client";

import React, { useEffect, useState } from "react";
import { useTranslations } from "next-intl";
import Skeleton from "@/app/components/skeleton";
import Link from "next/link";
import styles from "./index.module.scss";

export const WelcomeSection = () => {
  const t = useTranslations("HomePage");
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const timer = setTimeout(() => setIsLoading(false), 2000);
    return () => clearTimeout(timer);
  }, []);

  if (isLoading) {
    return (
      <section className={styles.welcome}>
        <Skeleton variant="medium" className={styles.title} />
        <Skeleton className={`${styles.description} ${styles.skeletonLarge}`} />
        <div className={styles.authButtons}>
          <Skeleton variant="small" className={styles.button} />
          <Skeleton variant="small" className={styles.button} />
        </div>
      </section>
    );
  }

  return (
    <section className={styles.welcome}>
      <h2 className={styles.title}>{t("title")}</h2>
      <p className={styles.description}>{t("description")}</p>
      <div className={styles.authButtons}>
        <Link href="/signin" className={styles.button}>
          {t("signIn")}
        </Link>
        <Link href="/signup" className={styles.button}>
          {t("signUp")}
        </Link>
      </div>
    </section>
  );
};
