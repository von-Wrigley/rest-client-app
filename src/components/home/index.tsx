"use client";

import React, { useEffect, useState } from "react";
import { useTranslations } from "next-intl";
import Skeleton from "@/app/components/skeleton";
import Link from "next/link";
import { useRouter } from "next/navigation";
import { supabase } from "@/helper/supabaseClient";
import { Session } from "@supabase/supabase-js";
import styles from "./index.module.scss";

export const WelcomeSection = () => {
  const t = useTranslations("HomePage");
  const router = useRouter();
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [session, setSession] = useState<Session | null>(null);

  useEffect(() => {
    const checkSession = async () => {
      const {
        data: { session },
      } = await supabase.auth.getSession();
      setSession(session);

      const timer = setTimeout(() => setIsLoading(false), 2000);
      return () => clearTimeout(timer);
    };

    checkSession();

    const {
      data: { subscription },
    } = supabase.auth.onAuthStateChange((_event, session) => {
      setSession(session);
    });

    return () => subscription.unsubscribe();
  }, []);

  const handleSignOut = async () => {
    await supabase.auth.signOut();
    router.refresh();
  };

  if (isLoading) {
    return (
      <section className={styles.welcome}>
        <Skeleton variant="medium" className={styles.title} />
        <Skeleton className={`${styles.description} ${styles.skeletonLarge}`} />
        <div className={styles.authButtons}>
          <Skeleton variant="small" className={styles.button} />
          <Skeleton variant="small" className={styles.button} />
          <Skeleton variant="small" className={styles.button} />
          <Skeleton variant="small" className={styles.button} />
        </div>
      </section>
    );
  }

  return (
    <section className={styles.welcome}>
      <h2 className={styles.title}>{t("title")}</h2>
      <p className={styles.description}>{t("description")}</p>

      <div className={styles.authButtons}>
        {session ? (
          <>
            <Link href="/collections" className={styles.button}>
              {t("collections")}
            </Link>
            <Link href="/history" className={styles.button}>
              {t("history")}
            </Link>
            <Link href="/variables" className={styles.button}>
              {t("variables")}
            </Link>
            <button onClick={handleSignOut} className={styles.button}>
              {t("signOut")}
            </button>
          </>
        ) : (
          <>
            <Link href="/signin" className={styles.button}>
              {t("signIn")}
            </Link>
            <Link href="/signup" className={styles.button}>
              {t("signUp")}
            </Link>
          </>
        )}
      </div>
    </section>
  );
};
