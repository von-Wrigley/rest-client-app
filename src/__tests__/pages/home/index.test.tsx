"use client";
import { render, screen, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { WelcomeSection } from "@/components/home";

jest.mock("next/link", () => {
  return ({ children, href }: { children: React.ReactNode; href: string }) => {
    return <a href={href}>{children}</a>;
  };
});

jest.mock("next-intl", () => ({
  useTranslations: () => (key: string) => key,
}));

describe("WelcomeSection component", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    act(() => {
      jest.runOnlyPendingTimers();
    });
    jest.useRealTimers();
  });

  it("renders correctly with all elements after loading", () => {
    render(<WelcomeSection />);
    act(() => {
      jest.advanceTimersByTime(2000);
    });

    expect(screen.getByRole("heading", { level: 2 })).toHaveTextContent(
      "title",
    );
    expect(screen.getByText("description")).toBeInTheDocument();

    const signInButton = screen.getByRole("link", { name: "signIn" });
    expect(signInButton).toHaveAttribute("href", "/signin");

    const signUpButton = screen.getByRole("link", { name: "signUp" });
    expect(signUpButton).toHaveAttribute("href", "/signup");
  });

  it("matches snapshot after loading", () => {
    const { asFragment } = render(<WelcomeSection />);
    act(() => {
      jest.advanceTimersByTime(2000);
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
