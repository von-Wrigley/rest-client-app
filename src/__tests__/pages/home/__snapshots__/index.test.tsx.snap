// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WelcomeSection component matches snapshot after loading 1`] = `
<DocumentFragment>
  <section
    class="welcome"
  >
    <h2
      class="title"
    >
      title
    </h2>
    <p
      class="description"
    >
      description
    </p>
    <div
      class="authButtons"
    >
      <a
        href="/signin"
      >
        signIn
      </a>
      <a
        href="/signup"
      >
        signUp
      </a>
    </div>
  </section>
</DocumentFragment>
`;
